types/src/params.rs

```diff
@@ -408,6 +416,256 @@ impl<'a> Id<'a> {
 	}
 }
 
+/// Initial number of bytes for a parameter length.
+const PARAM_BYTES_CAPACITY: usize = 128;
+
+/// Generic parameter builder that serializes parameters to bytes.
+/// This produces a JSON compatible String.
+///
+/// The implementation relies on `Vec<u8>` to hold the serialized
+/// parameters in memory for the following reasons:
+///   1. Other serialization methods than `serde_json::to_writer` would internally
+///      have an extra heap allocation for temporarily holding the value in memory.
+///   2. `io::Write` is not implemented for `String` required for serialization.
+#[derive(Debug)]
+struct ParamsBuilder {
+	bytes: Vec<u8>,
+	end: char,
+}
+
+impl ParamsBuilder {
+	/// Construct a new [`ParamsBuilder`] with custom start and end tokens.
+	/// The inserted values are wrapped by the _start_ and _end_ tokens.
+	fn new(start: char, end: char) -> Self {
```

Comment:
we could also get rid of the `fn new(start: char, end: char) -> Self`  here and provide two different constructs such as `fn positional() -> Self` and `fn named() -> Self `

For example

```rust
fn positional() -> Self {
    let mut bytes = Vec::with_capacity(PARAM_BYTES_CAPACITY);
    bytes.push(b'[');
    ParamsBuilder { bytes, end: b']' }
}
```


anything other than that isn't valid `JSON-RPC params` i.e, either `[ .... ]` or `{ .... }`

Reply:
It makes things explicit for us to provide the `positional` and `named` constructors.
I would still like to keep the `new` constructor, as we have just one part of the code to modify when things need changing. I made sure to make it private tho, and didn't think another file was worth it, so I also added it into a new module